### ISSUE TITLE
support for GPU-backed arrays

### ISSUE THREAD

**Issue by jrevels** [JuliaDiff/ReverseDiff.jl#44]
Originally, I thought we'd need a special macro directive for this, but I now believe this can just work by allowing the user to pass in GPU-backed arrays as input to the API methods. In other words, ReverseDiff won't provide a GPU-backed array type; it will compose with the GPU-backed array types provided by other well-established libraries. 

For now, [Julia's ArrayFire wrapper](https://github.com/JuliaComputing/ArrayFire.jl) should be sufficient to start playing around with this functionality for a subset of ReverseDiff's methods - at least enough to build some proof-of-concept examples. Eventually, I'd like to support [GPUArrays.jl](https://github.com/JuliaGPU/GPUArrays.jl), though it's probably not productive to work on that until GPUArrays has solid linear algebra coverage out-of-the-box.

In some cases, existing derivative definitions might "just work" for GPU-backed arrays (though they might have poor performance). In many cases, we'll likely have to dispatch on the array type to write special derivative methods for GPU-backed arrays.

cc the GPU-related folks: @SimonDanisch @vchuravy @maleadt @ranjanan @MikeInnes

**Comment by SimonDanisch** [comment]
Can we compile a list of operations that you need, to prioritize it in GPUArrays?

**Comment by jrevels** [comment]
> Can we compile a list of operations that you need, to prioritize it in GPUArrays?

Good idea. It would be nice to have a "master list" of supported TrackedArray functionality anyway.

**Comment by xukai92** [comment]
How is the current situation of ReverseDiff.jl working with GPUArrays.jl?

**Comment by SimonDanisch** [comment]
Different examples are working, but we haven't attempted to run a complete testsuite or anything ;)
Help to figure out what's covered and what needs improvement would be  appreciated!
